Move broadcast of key updates into sync, fixup of externaljwt generation / test

liggitt opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

1. Bugfix: Moves broadcast into syncKeys method, so any change to keys (periodic or on-demand) always notifies listeners immediately.
2. Test flake fix: Address race between setting new keys and ensuring they've been read (https://github.com/kubernetes/kubernetes/pull/128190#issuecomment-2462857707)
3. Add proto definition file needed for local generation (fixes https://github.com/kubernetes/kubernetes/issues/128672)



#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @enj
/sig auth

liggitt committed:
Move broadcast of key updates into sync

HarshalNeelkamal commented on pkg/serviceaccount/externaljwt/plugin/keycache.go:
```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
+	}
+	// If the order, key id, or oidc discovery flag changed, we changed.
+	for i := range oldPublicKeys.Keys {
+		if oldPublicKeys.Keys[i].KeyID != newPublicKeys.Keys[i].KeyID {
```
shall we short before doing this?

liggitt replied:
no, if the backend changes the key order, I'm fine propagating that as a change

enj commented on pkg/serviceaccount/externaljwt/plugin/keycache.go:
```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
+			p.broadcastUpdate()
+		}
+
 		return newPublicKeys, nil
```
```suggestion
		return nil, nil
```

liggitt replied:
fixed

enj commented on pkg/serviceaccount/externaljwt/plugin/keycache.go:
```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
+	if len(oldPublicKeys.Keys) != len(newPublicKeys.Keys) {
+		return true
+	}
+	// If the order, key id, or oidc discovery flag changed, we changed.
```
`bytes.Equal` compare the keys?

liggitt replied:
we don't have bytes at this point, we have opaque instantiated public keys, I wasn't sure about this bit... I think the data timestamp and key count / order / logical ids is fine

enj reviewed (commented):
Minor comments.

liggitt committed:
Deflake supported key ack

liggitt committed:
Add definitions for non-dockerized proto generation

liggitt commented:
addressed comments, also folded in distinct follow-up commits to deflake an integration flake @benluddy observed and make proto generation work locally

enj commented:
/lgtm
/approve
